drivers: fuelgauge: Add Onsemi LC709203F driver

DBS06 opened the pull request:
Add driver for the Onsemi LC709203F fuel gauge

kartben commented on dts/bindings/fuel-gauge/onnn,lc709203f.yaml:
```diff
@@ -0,0 +1,79 @@
+    enum:
+      - 0x0
+      - 0x1
+    default: 0x1
```
Please document default values
https://docs.zephyrproject.org/latest/build/dts/bindings-upstream.html#dt-bindings-default-rules

kartben commented on samples/fuel_gauge/lc709203f/README.rst:
```diff
@@ -0,0 +1,45 @@
+.. zephyr:code-sample:: max17048
+   :name: MAX17048 Li-Ion battery fuel gauge
```
You have copy-paste typos plus I'd rather you turn the existing fuel gauge sample into a generic one as there is nothing specific to either driver in it
I.e. have the sample expect a fuel_gauge0 alias or something like that

kartben replied:
`fuel_gauge_get_props` just seems to be a convenience wrapper around `fuel_gauge_get_prop` so I think it might be worth changing the code so that you "try" all the props?
Or maybe even better, since it's a sample meant to demonstrate how to use the API, showcase a mix of `fuel_gauge_get_props` and `fuel_gauge_get_prop` -- `fuel_gauge_get_props` for FUEL_GAUGE_RELATIVE_STATE_OF_CHARGE and FUEL_GAUGE_VOLTAGE (if your confident 100% of the fuel gauge will always have these) and then try `fuel_gauge_get_prop` one by one on some of the "bonus" properties?

DBS06 replied:
so actually similar to the tests. 
I think I will do a first iteration where it is the basic sample only with these, to have it at least working properly:
```
fuel_gauge_prop_t props[] = {
	FUEL_GAUGE_RELATIVE_STATE_OF_CHARGE,
	FUEL_GAUGE_VOLTAGE,
};
```
and next week or when I have time again, extend the sample as you suggested, because I like your idea very much.

kartben commented on boards/adafruit/feather_esp32s2/doc/adafruit_feather_esp32s2.rst:
```diff
@@ -311,14 +309,38 @@ There is a sample available to verify that the MAX17048 fuel gauge on the board
    :board: adafruit_feather_esp32s2@C
    :goals: build flash
 
+Testing the Fuel Gauge (LC709203F)
```
I would put everything under a "Testing the Fuel Gauge" section with two subsections

kartben reviewed (changes requested):
Some early feedback

kartben commented on drivers/fuel_gauge/lc709203f/emul_lc709203f.c:
```diff
@@ -0,0 +1,284 @@
+	uint16_t addr;
+};
+
+#define LC709203F_CRC_POLYNOMIAL 0x07 // Polynomial to calculate CRC-8-ATM
+
+/*
+ * Calculate the CRC-8 checksum using polynomial 0x07 (CRC-8-ATM).
+ */
+static uint8_t lc709203f_calc_crc(uint8_t *data, size_t data_len)
+{
+	uint8_t crc = 0;
+
+	for (size_t j = data_len; j; --j) {
+		crc ^= *data++;
+
+		for (size_t i = 8; i; --i) {
+			crc = (crc & 0x80) ? (crc << 1) ^ LC709203F_CRC_POLYNOMIAL : (crc << 1);
+		}
+	}
+
+	return crc;
+}
```
Use zephyr crc8 API?

DBS06 replied:
Didn't know that zephyr has it, but somehow it was clear, that this already exists :smirk:

kartben reviewed (commented):
Also please separate in multiple commits as it is doing more than just adding a new driver (i.e. adding new fuel gauge property typed)

DBS06 commented:
> Also please separate in multiple commits as it is doing more than just adding a new driver (i.e. adding new fuel gauge property typed



> Also please separate in multiple commits as it is doing more than just adding a new driver (i.e. adding new fuel gauge property typed)

I will split the commits asap

DBS06 commented:
> Some early feedback

Thanks a lot for this fast reply!

DBS06 committed:
drivers: fuelgauge: Add additional properties

Adds properties for current direction, charge and voltage alarms

Signed-off-by: Philipp Steiner <philipp.steiner1987@gmail.com>

DBS06 committed:
tests: fuelgauge: Add Onsemi LC709203F driver

Add test for the Onsemi LC709203F fuel gauge

Signed-off-by: Philipp Steiner <philipp.steiner1987@gmail.com>

DBS06 committed:
drivers: fuelgauge: format fuel_gauge.h

format fuel_gauge.h according to the clang-format settings

Signed-off-by: Philipp Steiner <philipp.steiner1987@gmail.com>

DBS06 committed:
drivers: fuelgauge: Add Onsemi LC709203F driver

Add driver for the Onsemi LC709203F fuel gauge

Signed-off-by: Philipp Steiner <philipp.steiner1987@gmail.com>

DBS06 commented:
> Also please separate in multiple commits as it is doing more than just adding a new driver (i.e. adding new fuel gauge property typed)

I split up the commit into multiple commits in a way which I think makes sense

DBS06 commented:
overview which is still missing or needs/should/wished to be changed

- [x] Copilot Review
- [x] Adapt Review
- [x] Change CRC calculation to Zephyr CRC
- [x] Add explanation for dts/bindings default values
- [x] Split up the commits
- [x] Add most basic fuelgauge sample
- [ ] Extend fuelgauge sample as discusses here: https://github.com/zephyrproject-rtos/zephyr/pull/89000#discussion_r2057643817

DBS06 committed:
boards: adafruit: esp32s2_feather: update documentation

update because of LC709203F driver was added

Signed-off-by: Philipp Steiner <philipp.steiner1987@gmail.com>

DBS06 committed:
samples: fuelgauge: unify fuel gauge sample

Replace IC specific fuel gauge example with a generic fuel gauge example

Signed-off-by: Philipp Steiner <philipp.steiner1987@gmail.com>

DBS06 committed:
drivers: fuelgauge: Add missing property values

Adds missing property values for battery connect and present state

Signed-off-by: Philipp Steiner <philipp.steiner1987@gmail.com>

DBS06 commented:
@kartben I also had to change my e-mail address for the copyright info, even for my last major PR #85405, because it was wrong. I am really sorry and somehow embarrassed for that, because actually this kind of mistakes makes me angry about myself because this does not meet my personal quality requirements!
Sometimes I shouldn't do things when I am tired, but with a little child at home the evenings are the only time to do such things :smirk: 
The only thing I am relieved about it, is that this mistakes didn't made it into a release.

teburd commented:
> @kartben @aaronemassey @teburd I have one questiong regarding the `doc/hardware/peripherals/fuel_gauge.rst` Imho I think this doc-file is partially outdated. Should I update it? Will the maintainers do it? I think I am already doing a lot of changes with the PR and to update this file is somehow part of the job.. If wished I could also open a new PR for this, but I think this is not necessary.

If you find missing/out of date docs then yes please send a separate PR updating them.

teburd commented:
Please review your commits and messages

The very first commit has added properties but the commit title is

"drivers: fuelgauge: Add Onsemi LC709203F driver"

This should be more like...

```
fuelgauge: Add additional properties

Adds properties for charge and voltage alarms

Signed-off-by: ...
```

With no formatting changes (which are there in the commit)

Basically the commits need to be cleaned up before I'll take a look again

teburd commented on tests/drivers/fuel_gauge/lc709203f/CMakeLists.txt:
```diff
@@ -0,0 +1,8 @@
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(device)
+
+FILE(GLOB app_sources src/test_lc709203f.c)
```
No need for a glob here for a single file, generally don't glob

teburd reviewed (commented):
Most of this looks really nice, please fix the commit message titles and messages, ensure format changes are separate, +1 from me on doing so

DBS06 commented:
> Most of this looks really nice, please fix the commit message titles and messages, ensure format changes are separate, +1 from me on doing so

Thanks a lot for your feedback, I really appreciate that!

DBS06 commented:
@teburd 

> > @kartben @aaronemassey @teburd I have one questiong regarding the `doc/hardware/peripherals/fuel_gauge.rst` Imho I think this doc-file is partially outdated. Should I update it? Will the maintainers do it? I think I am already doing a lot of changes with the PR and to update this file is somehow part of the job.. If wished I could also open a new PR for this, but I think this is not necessary.
> 
> If you find missing/out of date docs then yes please send a separate PR updating them.

I added a PR for this issue #89301

DBS06 commented:
> Most of this looks really nice, please fix the commit message titles and messages, ensure format changes are separate, +1 from me on doing so

I changed the commit messages, split the commits and reordered them. So far I think/hope I did this in the way as it was intended by you.

kartben commented:
> > Most of this looks really nice, please fix the commit message titles and messages, ensure format changes are separate, +1 from me on doing so
> 
> I changed the commit messages, split the commits and reordered them. So far I think/hope I did this in the way as it was intended by you.

@teburd can you please take another look? Thanks!

teburd reviewed (dismissed):
nice patch set

DBS06 committed:
Merge branch 'main' into add_LC709203F_fuel_gauge_driver

DBS06 committed:
Revert "Merge branch 'main' into add_LC709203F_fuel_gauge_driver"

This reverts commit ceeddb98a750b6b8e7eb88f35e77a4355b21034c, reversing
changes made to 03c4e8aa7a188540b6caf8984d7c234cc586d368.

DBS06 commented:
I messed up the commits due to GitHub's merge conflict solver, therefore I will close this PR and recreate the PR see #90310

DBS06 commented:
I messed up the commits due to GitHub's merge conflict solver, therefore I will close this PR and recreate the PR see https://github.com/zephyrproject-rtos/zephyr/pull/90310
